mne/preprocessing/maxwell.py

```diff
@@ -0,0 +1,255 @@
+    # Degree/order and theta/phi reversed
+    #return np.real(sph_harm(m, l, azimuth, polar))
+
+
+def get_num_harmonics(L_in, L_out):
+    """
+    Compute total number of spherical harmonics.
+
+    Parameters
+    ---------
+    L_in : int
+        Internal expansion order
+    L_out : int
+        External expansion order
+
+    Returns
+    -------
+    M : int
+        Total number of spherical harmonics
+    """
+
+    return L_in ** 2 + 2 * L_in + L_out ** 2 + 2 * L_out
+
+
+def _alegendre_deriv(l, m, x):
+    """
+    Compute the derivative of the associated Legendre polynomial at x.
+
+    Parameters
+    ----------
+    l : int
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    x : float
+        Value to evaluate the derivative at
+
+    Returns
+    -------
+    dPlm
+        Associated Legendre function derivative
+    """
+
+    #TODO: Eventually, probably want to switch to look up table but optimize
+    # later
+    C = 1
+    if m < 0:
+        m = abs(m)
+        C = (-1) ** m * fact(l - m) / fact(1 + m)
+    return C * (m * x * lpmv(m, l, x) + (l + m) * (l - m + 1) *
+                np.sqrt(1 - x ** 2) * lpmv(m, l - 1, x)) / (1 - x ** 2)
+
+
+def grad_in_comp(l, m, r, az, pol, lut_func):
+    """
+    Compute gradient of in-component of V(r) spherical expansion having form
+    Ylm(theta, phi) / (r ** (l+1))
+
+    Parameters
+    ----------
+    l : int
+        Degree of spherical harmonic
+    m : int
+        Order of spherical harmonic
+    r : numpy array of shape [n_samples,]
+        Array of radii
+    az: numpy array of shape [n_samples,]
+        Array of azimuthal (longitudinal) spherical coordinates [0, 2*pi]. 0 is
+        aligned with x-axis.
+    pol: numpy array of shape [n_samples,]
+        Array of polar (or colatitudinal) spherical coordinates [0, pi]. 0 is
+        aligned with z-axis.
+
+    Returns
+    -------
+    dPlm
+        Derivative of Associated Legendre function at points specified
+    """
+    #TODO: add check/warning if theta or phi outside appropriate ranges
+
+    # Get function for Legendre derivatives
+    #lut, n_fact = _get_legen_table('meg', False, 100)
+    #lut_deriv_fun = partial(_get_legen_lut_accurate, lut=lut)
+
+    # Compute gradients for r, theta, and phi
+    r1 = -(l + 1) / r ** (l + 2) * sph_harmonic(l, m, az, pol)
+
+    theta1 = 1 / r ** (l + 2) * np.sqrt((2 * l + 1) * factorial(l - m) /
+                                        (4 * np.pi * factorial(l + m))) * \
+        -np.sin(pol) * _alegendre_deriv(l, m, np.cos(pol)) * np.exp(1j * m *
+                                                                    az)
+
+    phi1 = 1 / (r ** (l + 2) * np.sin(pol)) * 1j * m * \
+        sph_harmonic(l, m, az, pol)
+
+    # Get real component of vectors, convert to cartesian coords, and return
+    return _to_real_and_cart(np.concatenate((r1, theta1, phi1)), m)
+
+
+def grad_out_comp(l, m, r, az, pol):
```

Comment:
avoid the four-letter `comp` because it is used in a lot of other places to mean `compensation` for CTF systems, so maybe `grad_out_components` or something